### DISCUSSION TITLE
Functions as a first class citizens

### DISCUSSION THREAD

**Discussion by igor-elovikov** [shader-slang/slang#5651] [Slang Proposed Features (SPFs)]
Consider a simple example of higher order function:
```c
import playground;

struct Add : IFunc<float, float, float> {
    float operator()(float arg0, float arg1) {
        return arg0 + arg1;
    }
}

float add_free_function(float arg0, float arg1) {
    return arg0 + arg1;
}

struct BindFront<F: IFunc<float, float, float>> : IFunc<float, float> {
    F fun;
    float value;

    float operator()(float arg) {
        return fun(arg, value);
    }
}

IFunc<float, float> bind_front<F: IFunc<float, float, float>>(F f, float value) {
    return BindFront<F>(f, value);
}

void printMain()
{
    let add = Add();
    printf("result: %g\n", add(1.0, 5.0)); // 6.0

    let add_five = bind_front(add, 5.0);
    printf("result: %g\n", add_five(1.0)); // 6.0


    // Below couldn't compile
    //let add_five_free = bind_front(add_free_function, 5.0); 
}
```
[Playground Link](https://shader-slang.com/slang-playground/?target=SPIRV&code=eJyFUs1KxDAQvucphgUxxdLtHvSwXQvuoeDBNxBKbdISyCYlm1RE-u5O0-pGLdtc8jMz389kxKnTxkInq4_WaKdYRsjZGldbeGIM9vBcOFUfGqkrG0O45fBJAJe_gO64qaw2NKLTQ2XadM4cz7toTh-X4dYZ5VPgzkczHxvIQMhcwljZGM7LBumt0GoFdgly-LFyFIoVRit7KK45yhf9fhstALVkgee-ko5jv653YcE44lCMxBNCFJhfYn9D8diLdfUUFcahtj_NCbqQ0-bCjry9Fgw6I5R9qYSiEZkKJfc_AY_jMNBZqE9r6Mbws5N2Dzftq9rEYx7dJfg590kaRRlst_CQpCTEKRvRcwS7OKL4OlWsYfvakSDE9jV4OXKp36HWTjJ1a_Fw6oTkczQk9zP1T8HvSZv1ABm-AFdU8_Q)

I think it would be better if a free function could conform to the `IFunc` interface, even if it requires introducing an anonymous type during specialization. Currently, `add_free_function` is of type `(float, float) -> float` according to the compilation error, but this type is not directly usable.

**Comment by jkwak-work** [comment]
`IFunc` is relatively new to Slang and you are discovering the limitations of the latest features.

For this specific case, the proposal [document](https://github.com/shader-slang/slang/blob/master/docs/proposals/009-ifunc.md) for `IFunc` already discussed as following,

> Eventually, we should allow ordinary functions to be automatically coerceable to IFunc interfaces. But this is scoped out for the initial IFunc work, because we believe the implementation can be simpler if we support lambda function first, then implement ordinary function coercion as a special case of lambda expressions.

I guess we will have lambda and the ordinary function coercion later.

**Comment by cheneym2** [comment]
Just to echo what Jay said, this is indeed a desirable feature but it is only informally planned.  I created https://github.com/shader-slang/slang/issues/5784 though I don't know when the feature will be added.
